spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,305 @@
+//                                   (i)    (i+1)    (i+2)
+func GenerateConnectionProofs(chains []*Chain) []*ProofData {
+    assert(len(chains) > 2)
+
+    var proofs []*ProofData
+
+    // iterate all but the last two chains
+    for i := 0; i < len(chains)-2; i++ {
+  
+        previousChain := chains[i]  // previous chains state root is on currentChain and is the source chain for i==0.
+        currentChain := chains[i+1] // currentChain is where the proof is queried and generated
+        nextChain := chains[i+2]    // nextChain holds the state root of the currentChain
+
+        currentHeight := currentChain.GetClientStateHeight(sourceChain) // height of previous chain state on current chain
+        nextHeight := nextChain.GetClientStateHeight(currentChain)      // height of current chain state on next chain
+
+        // prefixed key for the connection from currentChain to prevChain
+        connectionKey := GetPrefixedConnectionKey(currentChain)
+
+        // proof of current chain's connection to previous Chain.
+        connectionEnd := GetConnection(currentChain)
+
+        // Query proof of the currentChain's connection with the previousChain at nextHeight
+        // (currentChain's state root height on nextChain)
+        connectionProof := GetConnectionProof(currentChain, nextHeight)
+  
+        proofs = append(proofs, &ProofData{
+            Key:   &connectionKey,
+            Value: connectionEnd,
+            Proof: connectionProof,
+        })
+    }
+    return proofs       
+}
+```
+
+### Multi-hop Proof Verification Steps
+
+The followin outlines the general proof verification steps specific to a multi-hop IBC message.
+
+1. Unpack the multihop proof bytes into consensus, connection, and channel/commitment proof data.
+2. Iterate through connection proof data and verify each connectionEnd is in the OPEN state. Check connectionHops during channel handshake.
+3. Starting with known `ConsensusState[N-1]` at the given `proofHeight` on `ChainN` prove the prior chain's consensus and connection state.
+4. Repeat step 3, proving `ConsensusState[i-1]` and `Conn[i-1,i]` until `ConsensusState0` on the source chain is proven.
+   - Start with i = N-1
+   - ConsensusState <= ConsensusState[i-1] on Chain[i]
+   - ConsensusProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
+   - ConnectionProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
+   - Set ConsensusState from unmarshalled ConsensusProofs[i].Value
+5. Finally, prove the expected channel or packet commitment in `ConsensusState0`
+
+For more details see [ICS4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).
+
+### Multi-hop Proof Verification Pseudo Code
+
+Pseudocode proof verification of a channel between chains `A -> B -> C` .
+
+```typescript
+function verifyMembership(
```

Comment:
I think we need more explicit pseudocode here. Similar to what you have above in proof generation, where you loop through the consensus state and connection state proofs and verify them